### PR TITLE
Speed up uuidToPhysicalPath and uuidToObject.

### PR DESCRIPTION
**Pull request by mauritsvanrees**
Do this by using an `IndexQuery` to only query the UID index. This means we do not need to get a full brain, which takes long.

Before, in a `test_speed` run with 400 items:

```
Time taken for uuidToPhysicalPath: 0.21054577827453613
Time taken for uuidToObject: 0.4477109909057617
```

Afterwards:

```
Time taken for uuidToPhysicalPath: 0.030041933059692383
Time taken for uuidToObject: 0.09154772758483887
```

So this is 5 to 7 times as fast.

I tried changing `uuidToURL` along the same lines, but in both cases this was about the same, 0.15 seconds.

### CODE REVIEW

**Comment by mauritsvanrees** [comment]
This should work in Plone 5 and 6, but it may be better to only do this in Plone 6.

@jenkins-plone-org please run jobs

File: plone/app/uuid/utils.py [hunk 1]
```diff
@@ -1,18 +1,33 @@
...
+    if site is None:
         return None
-
-    return brain.getPath()
+    catalog = getToolByName(site, 'portal_catalog', None)
+    if catalog is None:
+        return None
+    index = catalog.Indexes["UID"]
+    # This works even faster, but uses a private attribute:
+    # rid = index._index.get(uuid)
```
Comment (jensens) [inline plone/app/uuid/utils.py]: IMO it is save to use `_index`.

File: plone/app/uuid/utils.py [hunk 2]
```diff
@@ -31,12 +46,13 @@ def uuidToObject(uuid):
...
-        return None
-
-    return brain.getObject()
+    path = uuidToPhysicalPath(uuid)
+    if not path:
+        return
+    site = getSite()
+    if site is None:
+        return
+    return site.restrictedTraverse(path)
```
Comment (jensens) [inline plone/app/uuid/utils.py]: I think here we need an `unrestrictedTraverse`, otherwise we run into permission problems if i.e. a  path `/a/b/c` with no permission for b but for a and c it will fail. Correct me if I am wrong. We may need a permission check for the resulting object then here.

File: plone/app/uuid/utils.py [hunk 3]
```diff
@@ -1,18 +1,33 @@
...
+    """
+    site = getSite()
+    if site is None:
         return None
-
-    return brain.getPath()
+    catalog = getToolByName(site, 'portal_catalog', None)
+    if catalog is None:
+        return None
+    index = catalog.Indexes["UID"]
```
Comment (jensens) [inline plone/app/uuid/utils.py:42]: Do we gain speed by caching the index i.e. on request? (which should not cause any harm)

**Review by jensens — COMMENTED** [review]
So, tests are not green. For now just a comment, but good finding!

File: plone/app/uuid/utils.py (hunk 2 above)
Comment (ale-rt) [reply plone/app/uuid/utils.py]: I would do what the `getObject` does.
It gets `a/b` unrestricted and the it does a `restrictedTraverse` to `c`, see:
https://github.com/zopefoundation/Products.ZCatalog/blob/d4fd9fe28d27e8ff43e911025c1258f1e8d50ad5/src/Products/ZCatalog/CatalogBrains.py#L88-L91

File: plone/app/uuid/utils.py (hunk 1 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py]: I try `_index` now, and fall back to `IndexQuery` when it fails.

File: plone/app/uuid/utils.py (hunk 3 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py:42]: I looked at caching the index on the request, but when you have a match for a uuid to an `rid`, then you still need to get the catalog to get the path for this `rid`.
So I now cache the portal_catalog on `request._catalog`. Seems fine, and indeed another speedup.

File: plone/app/uuid/utils.py (hunk 2 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py]: I switched to the combination of unrestricted and restricted traverse. I catch Unauthorized at the end, and return None then. This should be same as the original, where a catalog query would come up empty when you do not have permission on the object.

File: plone/app/uuid/utils.py (hunk 3 above)
Comment (jensens) [reply plone/app/uuid/utils.py:42]: Caching the catalog on request._catalog is probably a good idea overall in Plone (not only here). But that's a different topic.

File: plone/app/uuid/utils.py [hunk 4]
```diff
@@ -1,18 +1,56 @@
...
+def _catalog():
+    """Get and store portal_catalog on request.
+
+    This avoids looking up the site and the portal_catalog
+    each time one of the below functions is called.
+    """
+    request = getRequest()
+    try:
+        return request._catalog
+    except AttributeError:
```
Comment (jensens) [inline plone/app/uuid/utils.py:18]: AFAIK (correct me if I am wrong): A traceback is expensive. getattr + condition is way cheaper.

**Comment by mauritsvanrees** [comment]
I have addressed all comments. I have created a branch 2.0.x for Plone 5.2. So the 5.2 jobs can be ignored. (Closing and reopening the PR did not help and I did not want to create a new PR.)
Comparison with the original:

```
Time taken to create 40 items: 1.9684090614318848
Time taken for uuidToPhysicalPath orig: 0.015377044677734375
Time taken for uuidToPhysicalPath new:  0.0006089210510253906
Time taken for uuidToURL orig: 0.018133878707885742
Time taken for uuidToURL new:  0.015197992324829102
Time taken for uuidToObject orig: 0.020327091217041016
Time taken for uuidToObject new:  0.005269050598144531
Time taken for uuidToCatalogBrain orig: 0.014737844467163086
Time taken for uuidToCatalogBrain new:  0.011925935745239258
```

With a larger run:

```
Time taken to create 1000 items: 58.546963930130005
Time taken for uuidToPhysicalPath orig: 0.3945882320404053
Time taken for uuidToPhysicalPath new:  0.010138988494873047
Time taken for uuidToURL orig: 0.380126953125
Time taken for uuidToURL new:  0.33528780937194824
Time taken for uuidToObject orig: 0.617110013961792
Time taken for uuidToObject new:  0.1368708610534668
Time taken for uuidToCatalogBrain orig: 0.3616199493408203
Time taken for uuidToCatalogBrain new:  0.2982752323150635
```

All new versions are faster. Biggest win is in `uuidToPhysicalPath` which is 25 to 40 times faster. But this function is hardly used outside of `plone.app.uuid`.
The most relevant comparison is for `uuidToObject`, which is the only function that is used in more than two core packages. The new version is about 4 times faster.

@jenkins-plone-org please run jobs

File: plone/app/uuid/utils.py (hunk 4 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py:18]: I think trying and catching an exception is usually cheaper. Let's test it.  I tried getting the catalog one million times. With the above exception handling: 0.53 seconds.
With gettattr: at first 0.56 seconds. On second try 0.52.  So seems about the same.
This was the code:

```
    # _marker = object() is defined outside of the function.
    catalog = getattr(request, '_catalog', _marker)
    if catalog is not _marker:
        return catalog
```

File: plone/app/uuid/utils.py [hunk 5]
```diff
@@ -31,26 +72,30 @@ def uuidToObject(uuid):
...
+    # Go to the parent of the item without restrictions.
+    split_path = path.split("/")
+    parent_path = split_path[:-1]
+    parent = site.unrestrictedTraverse(parent_path)
+    # Do check restrictions for the final object.
+    final_path = split_path[-1]
+    try:
+        return parent.restrictedTraverse(final_path)
+    except Unauthorized:
+        return
```
Comment (ale-rt) [inline plone/app/uuid/utils.py]: Is there a reason why you want to change the previous implementation and returning `None` even if before it was raising the exception?

See the previous behavior in action:
```
In : from zope.component.hooks import setSite
In : from plone.app.uuid.utils import uuidToObject
In : setSite(app.Plone)
In : with api.env.adopt_roles(roles=["foo"]): uuidToObject(app.Plone.portal_catalog.unrestrictedSearchResults(review_state="private")[0].UID)
...: 
Traceback (most recent call last):
...
AccessControl.unauthorized.Unauthorized: You are not allowed to access ...
```

My suggestion is
```suggestion
    parent_path, final_path = path.rpartition("/")[::2]
    # Go to the parent of the item without restrictions,
    # but do a security check for the final object, 
    # like CatalogBrain.getObject does
    parent = site.unrestrictedTraverse(parent_path)
    return parent.restrictedTraverse(final_path)
```

File: plone/app/uuid/tests.py [hunk 6]
```diff
@@ -59,6 +61,58 @@ def test_uuidToPhysicalPath(self):
...
+            }
+        end = time.time()
+        print("Time taken to create {} items: {}".format(total, end - start))
+
+        self.assertEqual(len(uuids), total)
+        start = time.time()
+        for uuid, info in uuids.items():
+            self.assertEqual(info['path'], uuidToPhysicalPath(uuid))
+        end = time.time()
+        print("Time taken for uuidToPhysicalPath: {}".format(end - start))
```
Comment (ale-rt) [inline plone/app/uuid/tests.py]: It is pretty unusual for the tests to print some output, see https://jenkins.plone.org/job/pull-request-6.0-3.8/756/consoleFull

File: plone/app/uuid/utils.py (hunk 5 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py]: I did not realise the previous code was getting unrestrictedSearchResults from the catalog. I would argue that the previous code is wrong then. Well, not helpful anyway. Always getting None back when you can't get an object, seems easier than handling both None and a possible Unauthorized exception.

So I would be inclined to keep my code.
But then it is no longer only about improving the speed, so perhaps that part is best done in a separate PR...

Comment (ale-rt) [reply plone/app/uuid/utils.py]: I see your point and I am fine with the breaking change fix :)
Go ahead if you want!

File: plone/app/uuid/tests.py (hunk 6 above)
Comment (mauritsvanrees) [reply plone/app/uuid/tests.py]: I have changed it to only print when an environment variable is set:

```
$ PLONE_APP_UUID_TEST_SPEED_TOTAL=100 bin/test -s plone.app.uuid -t test_speed
...
Creating 100 documents...
Time taken to create 100 items: 4.955342054367065
Time taken for uuidToPhysicalPath: 0.0012142658233642578
Time taken for uuidToURL: 0.05345273017883301
Time taken for uuidToObject: 0.02753305435180664
Time taken for uuidToCatalogBrain: 0.030633926391601562
```

File: plone/app/uuid/utils.py (hunk 5 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py]: I have added more tests to check that we return None in case we have no permission.
I  have changed `uuidToCatalogBrain` along the same lines: call `searchResults` instead of `unrestrictedSearchResults`. `uuidToURL` calls this function, so it has the new logic as well.

**Comment by mauritsvanrees** [comment]
One `plone.app.linkintegrity` test fails. It tests an event handler that is fired when an object is modified. This drops all link integrity relations, and creates new relations for all the uuid references it finds. Problem is that an Editor may not have access to all objects that the current item links to. With my new code, this means that those objects are not found, so you lose some relations. When a Manager edits the item, the relations are put back again.
That is not good.

I will revert to the previous behavior.

We could think about adding an extra keyword argument to most of these functions: `restricted/unrestricted=True/False`. Question is then which name to use, and what the default would be. Best done in a different PR.

File: plone/app/uuid/utils.py [hunk 7]
```diff
@@ -30,27 +74,28 @@ def uuidToURL(uuid):
...
+    Note: the user may not be authorized to view the object.
+    It is up to the caller to check this, if needed.
+    """
+    path = uuidToPhysicalPath(uuid)
+    if not path:
+        return
+    site = getSite()
+    if site is None:
+        return
+    return site.unrestrictedTraverse(path)
```
Comment (ale-rt) [inline plone/app/uuid/utils.py]: This is again a breaking change... What happened to the code that was getting the parent unrestricted and the final path restricted?

**Review by ale-rt — APPROVED** [review]
I am sorry this is costing you so much effort.
If you want merge it and I will take care of adding the `(un)restricted` keyword parameter

File: plone/app/uuid/utils.py (hunk 7 above)
Comment (mauritsvanrees) [reply plone/app/uuid/utils.py]: Ah, right, the original code called `getObject`, and this does check security.
Okay, I have restored getting the parent unrestricted but the final path restricted, possibly raising Unauthorized.

Comment (mauritsvanrees) [reply plone/app/uuid/utils.py]: I am glad that you pay attention! :-)

**Review by ale-rt — APPROVED** [review]
